### ISSUE TITLE
MetricFilterAutoConfiguration doesn't match registered uriTemplates if login is required

### ISSUE THREAD

**Issue by bvoss** [spring-projects/spring-boot#2563]
Scenario: controller registered under `/document/{docId}/show`
Metrics are measured with the name `document.docid.show` -> great :-)

But if you add security and set the url to need a fully authenticated user, there might be (no there will be a lot!) unauthenticated requests to this resource which leads to redirects to the login page. So far so good, not really a problem you would say. Well it is because in this case nobody can set the `BEST_MATCHING_PATTERN_ATTRIBUTE` to get a grouping for the docIds.

As consequence your metrics mess with names like

```
document.123.show 
document.124.show
.
.
document.987456.show
```

:-(

**Comment by bvoss** [comment]
It would be great to have the grouping in this case as well, but I think this is hard to achieve.

But what about a special metric like 'unmapped' which prevents this metric mess for unmapped 404 requests?

My suggestion is to have 'authentication_required' and 'authorization_insufficient' metric, when the security filter chain stops the request

**Comment by dsyer** [comment]
This is pretty nasty and I feel like we should prioritise doing something for 1.2.6. I'm not really sure what yet. Maybe make 403 and 401 into special cases, and record those without the full URL path if we can't find the best match?

People using Jersey, or other non-Spring MVC frameworks, probably suffer in the same way, so I think we also need a better strategy longer term.

**Comment by bvoss** [comment]
Just a suggestion

Introduce some kind of interface 'RequestedUrlToMetricNameStrategy'. With one method 'String metricNameFor(HttpServletRequest request)'

With 3 implementations provided:
1 SpringMvcBestMatchingPattern
2 SpringSecurityNotAllowed
3 UriPatternMatch

The MetricFilter should ask all Beans of this interface in the application context until the first returned none null.

And finally do a fallback (which my be disabled) to something like 'unmapped'. 

Implementation 1 is in place by defaulte. 2 in case of SpringSecurity configured.
3 Just for easy use in none SpringMvc frameworks.

**Comment by dsyer** [comment]
I took a look at this and there is already logic that puts all unmapped 4xx errors in the same metric (it's been there for aeons as far as I can tell from a brief search through the history). Maybe @bvoss could provide a sample project if he is seeing this in an actual app?

**Comment by bvoss** [comment]
I can provide an example (this will take a while), but let me say, that my actual problem isn't solved, because I get no 4xx response but 302 to the login page / authentication service.

**Comment by dsyer** [comment]
Ah, it's a 302. That's completely different. If your metric names had included the 302 status in your original post I would have spotted that. Thinking...

**Comment by dsyer** [comment]
Fixed simply in 1.2.6. Possibly for 1.3 we should consider adding *all* unmapped requests to the same metric to avoid issues with non Spring MVC frameworks.

**Comment by bvoss** [comment]
Great!

and
+1 for all to unmapped
